### PR TITLE
crontab schedule

### PR DESCRIPTION
**Pull request by adam-fowler**
Add support for crontab style schedules
```swift
let jobSchedule = JobSchedule([
    (job: RemoveDeadSessionsJob(), schedule: .crontab("0 4 * * 7")), // four in the morning on Sunday
    (job: BirthdayRemindersJob(), schedule: .crontab("0 9,16 * * 1-5")), // nine in the morning and four in the afternoon on week days
])
```
Syntax supported
  `*` - wildcard
  `*/n` - repeating value every so many units
  `a-b` range of values
  `a-b/n` - repeating value every so many units within range
  `a,b,c...` selection of values

If both date and week day are both set we can't support one of them being a range of values.
Currently don't support dealing with dates that go past the end of the month eg 30th Feb.  This should be fixable

### CODE REVIEW

File: Sources/Jobs/Scheduler/Schedule+crontab.swift [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+    /// - Parameters:
+    ///   - crontab: Crontab string
+    ///   - timeZone: Timezone to run schedule in
+    /// - Throws: ScheduleError for corrupt crontabs and crontabs we don't suppoty
+    static func crontab(_ crontab: String, timeZone: TimeZone = .current) throws -> Self {
+        let values = crontab.split(separator: " ", omittingEmptySubsequences: true)
+        guard values.count == 5 else { throw ScheduleError("Crontab string requires 5 values") }
+        let minutes = try Self.parse(values[0], range: 0...60) { $0 }
+        let hours = try Self.parse(values[1], range: 0...24) { $0 }
+        let date = try Self.parse(values[2], range: 0...31) { $0 }
```
Comment (thoven87) [inline Sources/Jobs/Scheduler/Schedule+crontab.swift:34]: Should we keep the name as in cron syntax? i.e this variable should be called `dayOftheMonth`

File: Sources/Jobs/Scheduler/Schedule+crontab.swift [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        let month = try Self.parse(values[3], range: 1...12) {
+            guard let month = Month(rawValue: $0) else { throw ScheduleError("Invalid month value") }
+            return month
+        }
+        let day = try Self.parse(values[4], range: 0...6) {
```
Comment (thoven87) [inline Sources/Jobs/Scheduler/Schedule+crontab.swift:39]: here, this would be called `dayOftheWeek`?

File: Sources/Jobs/Scheduler/Schedule+crontab.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                    ","
+                }
+                OneOrMore(.digit)
+            } transform: {
+                $0.split(separator: ",").map { Int($0)! }
+            }
+            Anchor.endOfLine
+        }
+        if entry == "*" { return .any }
+        if let values = try? numberRegex.wholeMatch(in: entry) {
```
Comment (thoven87) [inline Sources/Jobs/Scheduler/Schedule+crontab.swift:111]: is the exception not swallowed here?

**Comment by thoven87** [comment]
This is great @adam-fowler the only thing I would are the followings

`@yearly` or `@annually`
`@monthly`
`@hourly`
`@daily` or `@midnight`
`@hourly`

I think the above presets are common enough to warrant mappings

**Comment by adam-fowler** [comment]
> This is great @adam-fowler the only thing I would are the followings
> 
> 
> 
> `@yearly` or `@annually`
> 
> `@monthly`
> 
> `@hourly`
> 
> `@daily` or `@midnight`
> 
> `@hourly`
> 
> 
> 
> I think the above presets are common enough to warrant mappings 

Can't you use the functions that are already available for those. Eg `.hourly()`, `.daily(hour: 4)`

File: Sources/Jobs/Scheduler/Schedule.swift [hunk 4]
```diff
@@ -291,15 +388,14 @@ public struct Schedule: Sendable {
...
     }
 
-    mutating func moveToPreviousScheduledDate() {
-        _ = self.second.prevValue()
-        _ = self.minute.prevValue()
-        _ = self.hour.prevValue()
-        _ = self.day.prevValue()
-        _ = self.date.prevValue()
-        _ = self.month.prevValue()
+    private static let daysInMonth = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]
```
Comment (thoven87) [inline Sources/Jobs/Scheduler/Schedule.swift:394]: Is there a specific reason the days are defined as above?

File: Tests/JobsTests/JobSchedulerTests.swift [hunk 5]
```diff
@@ -357,6 +357,63 @@ final class JobSchedulerTests: XCTestCase {
...
+            lastScheduled: "2023-02-03T05:00:00Z",
+            now: "2023-02-03T05:29:00Z",
+            expected: ["2023-02-27T10:00:00Z", "2023-02-28T10:00:00Z", "2023-03-27T10:00:00Z", "2023-03-28T10:00:00Z"]
+        )
+        try testScheduleAfterReadingLastData(
+            schedule: .crontab("0 10 27-30 * *", timeZone: .init(secondsFromGMT: 0)!),
+            accuracy: .latest,
+            lastScheduled: "2024-02-03T05:00:00Z",
+            now: "2024-02-03T05:29:00Z",
+            expected: ["2024-02-27T10:00:00Z", "2024-02-28T10:00:00Z", "2024-02-29T10:00:00Z", "2024-03-27T10:00:00Z"]
```
Comment (thoven87) [inline Tests/JobsTests/JobSchedulerTests.swift:413]: The expected values here seem a little bit funny. I would have expected, `2024-03-28T10:00:00Z`, `2024-03-29T10:00:00Z` and `2024-03-30T10:00:00Z` in the expected values as well.

**Comment by adam-fowler** [comment]
I personally believe adding in multiple APIs that do the same thing is bad API design. Also the `@yearly`, `@monthly` etc do not define when in the year, month they trigger.

**Comment by thoven87** [comment]
> I personally believe adding in multiple APIs that do the same thing is bad API design. Also the `@yearly`, `@monthly` etc do not define when in the year, month they trigger.

I understand your point of view Adam. It's already mutually agree that if you choose to pass in `@yearly`, it'll default to the first day of the year at midnight UTC. Same is true for `@monthly`, first day of the month at midnight UTC. If a user needs specific, the `.monthly(hour:Date)` e.t.c API can be used.

File: Sources/Jobs/Scheduler/Schedule.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/Jobs/Scheduler/Schedule.swift:415]: Not sure what you mean here. I need the day in each month so I can tell if the date has overflowed the number of days in the current month eg 31st April

File: Tests/JobsTests/JobSchedulerTests.swift (hunk 5 above)
Comment (adam-fowler) [reply Tests/JobsTests/JobSchedulerTests.swift:413]: I can add them in if you want, but the test is verifying what I wanted to verify ie we skip any out of range dates.

**Comment by adam-fowler** [comment]
> I understand your point of view Adam. It's already mutually agree that if you choose to pass in `@yearly`, it'll default to the first day of the year at midnight UTC. Same is true for `@monthly`, first day of the month at midnight UTC. If a user needs specific, the `.monthly(hour:Date)` e.t.c API can be used.

I really not keen on these. But I've added the nicknames found on the linux man page (does not include midnight). If you can find a link documenting midnight I can add.

File: Tests/JobsTests/JobSchedulerTests.swift (hunk 5 above)
Comment (thoven87) [reply Tests/JobsTests/JobSchedulerTests.swift:413]: The name of the variable is what's throwing me off. I see expected, which makes me believe that's the available range from min - max and anything not defined is not expected.

File: Sources/Jobs/Scheduler/Schedule.swift (hunk 4 above)
Comment (thoven87) [reply Sources/Jobs/Scheduler/Schedule.swift:415]: I thought the Calendar API had helpers which could help here, I suppose that's not the case.

**Comment by thoven87** [comment]
> > I understand your point of view Adam. It's already mutually agree that if you choose to pass in `@yearly`, it'll default to the first day of the year at midnight UTC. Same is true for `@monthly`, first day of the month at midnight UTC. If a user needs specific, the `.monthly(hour:Date)` e.t.c API can be used.
> 
> I really not keen on these. But I've added the nicknames found on the linux man page (does not include midnight). If you can find a link documenting midnight I can add.

The `0` in the syntax indicates that it's the first hour of the day which is midnight in the Gregorian calendar. Minute, hour, week are all zero based (minute is 0-59, hour is 0-23 and week is 0-6 | 0-7 where 7 is again Sunday ) If a link is still needed, I suppose either one of these should suffice 

https://en.wikipedia.org/wiki/Cron

https://man7.org/linux/man-pages/man5/crontab.5.html#top_of_page if we look at the "EXAMPLE CRON FILE" we can see this 

```
CRON_TZ=Japan
# run five minutes after midnight, every day
5 0 * * *       $HOME/bin/daily.job >> $HOME/tmp/out 2>&1
```

File: Sources/Jobs/Scheduler/Schedule.swift (hunk 4 above)
Comment (adam-fowler) [reply Sources/Jobs/Scheduler/Schedule.swift:415]: I can't find any

File: Tests/JobsTests/ScheduleTests.swift [hunk 6]
```diff
@@ -182,4 +182,23 @@ final class ScheduleTests: XCTestCase {
...
+            Schedule(minute: 0, hour: 0, date: .any, month: .any, day: .any)
+        )
+        XCTAssertEqual(
+            try Schedule.crontab("@monthly"),
+            Schedule(minute: 0, hour: 0, date: 1, month: .any, day: .any)
+        )
+        XCTAssertEqual(
+            try Schedule.crontab("@yearly"),
+            Schedule(minute: 0, hour: 0, date: 1, month: .init(.january), day: .any)
+        )
```
Comment (thoven87) [inline Tests/JobsTests/ScheduleTests.swift:201]: `@weekly` and `@annually` are missing :(

File: Tests/JobsTests/JobSchedulerTests.swift (hunk 5 above)
Comment (thoven87) [reply Tests/JobsTests/JobSchedulerTests.swift:413]: Understood! I will resolve this thread, thanks!
